packages/payment-widget/src/lib/utils/request.ts

```diff
@@ -0,0 +1,244 @@
+    const chain = getChainFromNetwork(targetNetwork);
+    if (!chain) {
+      throw new Error(`Unsupported network: ${targetNetwork}`);
+    }
+    targetChain = chain;
+  };
+
+  const ensureCorrectNetwork = async () => {
+    const currentChainId = await ethersProvider
+      .getNetwork()
+      .then((net) => net.chainId);
+
+    if (currentChainId !== targetChain.chainId) {
+      try {
+        await ethersProvider.send("wallet_switchEthereumChain", [
+          { chainId: utils.hexValue(targetChain.chainId) },
+        ]);
+      } catch (switchError: any) {
+        if (switchError.code === 4902) {
+          try {
+            await ethersProvider.send("wallet_addEthereumChain", [
+              getNetworkParams(targetChain),
+            ]);
+          } catch (addError) {
+            throw new Error(
+              `Failed to add and switch to the required network: ${targetChain.name}`
+            );
+          }
+        } else {
+          throw new Error(
+            `Failed to switch to the required network: ${targetChain.name}`
+          );
+        }
+      }
+      await initializeProvider();
+    }
+  };
+
+  const isERC20 =
+    requestParameters.requestInfo.currency.type ===
+    Types.RequestLogic.CURRENCY.ERC20;
+
+  await initializeProvider();
+  await ensureCorrectNetwork();
+
+  const web3SignatureProvider = new Web3SignatureProvider(
+    ethersProvider!.provider
+  );
+
+  const inMemoryRequestNetwork = new RequestNetwork({
+    nodeConnectionConfig: {
+      baseURL: "https://gnosis.gateway.request.network",
+    },
+    signatureProvider: web3SignatureProvider,
+    skipPersistence: true,
+  });
+
+  const inMemoryRequest =
+    await inMemoryRequestNetwork.createRequest(requestParameters);
+
+  const signer = await ethersProvider!.getSigner();
+  if (isERC20) {
+    const requestData = inMemoryRequest.inMemoryInfo?.requestData!;
+    const tokenAddress = requestData.currencyInfo.value;
+    const amount = requestData.expectedAmount;
+
+    const erc20Contract = new Contract(
+      tokenAddress,
+      ["function allowance(address,address) view returns (uint256)"],
+      signer
+    );
+
+    const _hasSufficientFunds = await hasSufficientFunds({
+      request: inMemoryRequest.inMemoryInfo?.requestData!,
+      address: payerAddress,
+      providerOptions: {
+        provider: ethersProvider!,
+      },
+    });
+
+    if (!_hasSufficientFunds) {
+      throw new Error("Insufficient funds");
+    }
+
+    const currentAllowance = await erc20Contract.allowance(
+      payerAddress,
+      payerAddress
+    );
+
+    if (currentAllowance.lt(amount)) {
+      const _approve = await approveErc20(
+        inMemoryRequest.inMemoryInfo?.requestData!,
+        signer
+      );
+      await _approve.wait(1);
+    }
+  }
+
+  const paymentTx = await payRequest(
+    inMemoryRequest.inMemoryInfo?.requestData!,
+    signer
+  );
+
+  await paymentTx.wait(1);
```

Comment:
The number of required block confirmations will differ depending on the chain - fast chains like Polygon require more block confirmations. Recommend 15 for Polygon.